### ISSUE TITLE
Formatting logic seems to be wrong when adding <p>

### ISSUE THREAD

**Issue by JohnKuan** [stevengharris/MarkupEditor#8]
Hi @stevengharris,

For the formatting logic (Bold, Italics, Underline, etc), the insertion of newline "p" appears to break the formatting structure of previous text. This is especially apparent when the structure becomes nested with permutations of B/I/U and P tags.
I am comparing this behaviour with email clients like outlook and gmail which provides similar formatting functionalities and it does not feel like this matches the behaviour of those email clients.

I have attached a screen recording of this occurrence.
https://user-images.githubusercontent.com/4402214/167060824-cecd4578-36df-4932-9813-d0f4be8f114e.mp4

Do let me know if this is an intended behaviour for formatting the text, and also do let me know if you require additional assistance on this.

**Comment by JohnKuan** [comment]
Hi @stevengharris, let me try to upload this screen recording again.
https://user-images.githubusercontent.com/4402214/167326420-3ddbd342-7943-4b48-9854-43ff5856c11b.mp4




> Not sure why I can't see your recording, but the intended behavior should match your expectations from email clients, etc. In my local branch this works fine (altho maybe you are seeing something different that would be clearer if I could see your screen recording). I have been working on the "multiple selection" logic and in doing so have dealt with some bugs in nested "format elements", which is what I've been calling `<B>`, `<I>`, etc.
> 
> If you insert a paragraph in the demo.html that looks like `<p><b>Hello <u id="u">bold and underline</u> world</b></p>` and you select inside `an|d` where `|` is the selection point, you should see `<p><b>Hello <u id="u">bold an</u></b></p><p><b><u>d underline</u> world</b></p>`. I have tests now to make sure this type of thing works.

**Comment by JohnKuan** [comment]
I tried with the paragraph `<p><b>Hello <u id="u">bold and underline</u> world</b></p>` in demo.html like you have mentioned and it does perform fine when I just entered a new line to it. 

However let's suppose if the user is composing a sentence with formatting as he types, the logic seems to break at that point.

I have attached another recording to illustrate my point. 
https://user-images.githubusercontent.com/4402214/167328074-15d612ea-2a49-41a5-9823-0f579f25161b.mp4

These are the visible observations in the video
- Formatting can be toggled off when adding italics and new line (underlined got removed)
- Removing a formatting by manually turning off with the button, can remove all of the previous text formatting. (Removing bold)

**Comment by stevengharris** [comment]
Thanks for the details. I managed to see the recording also, by using Chrome. Not sure why Safari doesn't work for them.

These issues should be fixed in https://github.com/stevengharris/MarkupEditor/commit/fe4ecb0ff28a4aee518c2d4e58084fd7751bc135 and https://github.com/stevengharris/MarkupEditor/commit/5d7ae155c2303750526445d52c8f35a343a9f585 .

I tagged version 0.3.2 if you want to just use it. I am going to switch to PRs for my own work in case it makes future fixes easier. Appreciate the feedback, which is timely as I try to close in on feature-complete and finish clean-up.
